external-crates/move/crates/move-vm-runtime/src/interpreter/eval.rs

```diff
@@ -0,0 +1,1158 @@
+
+    let mut run_context = RunContext {
+        loader,
+        link_context,
+        defn_resolver: resolver,
+        extensions,
+    };
+
+    let mut state = start_state;
+
+    let mut buf = String::new();
+    if DEBUG_STEP_PRINT {
+        let _ = state.debug_print_stack_trace(&mut buf, loader);
+        println!("Call Frame:\n{:?}", state.current_frame);
+        println!("{buf}");
+        buf.clear();
+    }
+
+    // Run until we're done or we produce an error and bail.
+    while step(&mut state, &mut run_context, gas_meter)? != StepStatus::Done {
+        if DEBUG_STEP_PRINT {
+            println!("-------------------------------------");
+            println!("Call Frame:\n{:?}", state.current_frame);
+            let _ = state.debug_print_stack_trace(&mut buf, loader);
+            println!("{buf}");
+            buf.clear();
+        }
+        continue;
+    }
+
+    // When we are done, grab the operand stack as the return type.
+    let MachineState { operand_stack, .. } = state;
+    let RunContext { .. } = run_context;
+    Ok(operand_stack.value)
+}
+
+fn step(
+    state: &mut MachineState,
+    run_context: &mut RunContext,
+    gas_meter: &mut impl GasMeter,
+) -> VMResult<StepStatus> {
+    let fun_ref = state.current_frame.function();
+    let instructions = fun_ref.code();
+    let pc = state.current_frame.pc as usize;
+    assert!(pc <= instructions.len());
+    let instruction = &instructions[pc];
+
+    trace!(
+        fun_ref,
+        &state.current_frame.locals,
+        state.current_frame.pc,
+        instruction,
+        run_context.defn_resolver,
+        state
+    );
+
+    fail_point!("move_vm::interpreter_loop", |_| {
+        Err(state.set_location(
+            PartialVMError::new(StatusCode::VERIFIER_INVARIANT_VIOLATION)
+                .with_message("Injected move_vm::interpreter verifier failure".to_owned()),
+        ))
+    });
+
+    profile_open_instr!(gas_meter, format!("{:?}", instruction));
+    if DEBUG_STEP_PRINT {
+        println!("Instruction: {instruction:?}");
+    }
+    // These are split out because `PartialVMError` and `VMError` are different types. It's unclear
+    // why as the hold identical data, but if we could combine them, we could entirely inline
+    // `op_step` into this function.
+    match instruction {
+        Bytecode::Ret => {
+            let charge_result = gas_meter.charge_simple_instr(SimpleInstruction::Ret);
+            profile_close_instr!(gas_meter, format!("{:?}", instruction));
+
+            partial_error_to_error(state, run_context, charge_result)?;
+            let non_ref_vals = state
+                .current_frame
+                .locals
+                .drop_all_values()
+                .map(|(_idx, val)| val);
+
+            // TODO: Check if the error location is set correctly.
+            gas_meter
+                .charge_drop_frame(non_ref_vals.into_iter())
+                .map_err(|e| state.set_location(e))?;
```

Comment:
this may very well be a "fix all things gas" issue, but I do not believe we charge anything for dropping a frame (lines 162-171) and yet we got calls and check and a TODO.... 
the`drop_all_values` part may actually do something useful (maybe not sure) but...

Reply:
This TODO was there in the old code -- not sure what to do about it. Let's revisit gas in the future.